### ISSUE TITLE
Custom node (sometimes) breaks when added to the flow

### ISSUE THREAD

**Issue by computerFriend** [node-red/node-red#1758]
### To Reproduce:
- (arguably optional): Develop a node locally & then publish it to npm
- install the custom node into your NodeRed project (either locally or from npm; same behavior is observed)
- start NodeRed locally
- add the custom node to a flow

### What happens?
- The node looks fine in the palette, with the info tab populated:
<img width="1525" alt="screen shot 2018-06-08 at 6 43 54 pm" src="https://user-images.githubusercontent.com/25398257/41186074-eb19c686-6b5d-11e8-8dd2-26b945e3be02.png">

- When that node is dragged onto a flow, the following errors are observed:
- The info tab for that node is suddenly blank
- The node will not open
- If the changes are deployed, the node shows up with blank attributes in the flows.json file: only position coordinates, id, & ret have values; all others have empty strings as values
- The console log shows the following stack trace:

> TypeError: Cannot read property 'module' of undefined
>     at Array.<anonymous> (red.js:15395)
>     at Object.emit (red.js:55)
>     at Object.addNode [as add] (red.js:2535)
>     at HTMLDivElement.drop (red.js:11019)
>     at e.(:443/anonymous function).(anonymous function)._trigger (http://localhost:443/vendor/vendor.js:16:10003)
>     at e.(:443/anonymous function).(anonymous function)._drop (http://localhost:443/vendor/vendor.js:16:40380)
>     at e.(:443/anonymous function).(anonymous function)._drop (http://localhost:443/vendor/vendor.js:16:5029)
>     at e.(:443/anonymous function).(anonymous function).<anonymous> (http://localhost:443/vendor/vendor.js:16:42279)
>     at Function.each (vendor.js:2)
>     at Object.drop (vendor.js:16)

<img width="1522" alt="screen shot 2018-06-08 at 8 54 09 pm" src="https://user-images.githubusercontent.com/25398257/41186086-2105f56c-6b5e-11e8-8ca5-da6d3cc63630.png">

Furthermore, this behavior is _inconsistent_, in that, sometimes (with no alteration to the custom node code) - this issue will vanish altogether, & the node will work fine.  Changes in behavior usually result from uninstalling & reinstalling the node.

### What do you expect to happen?
The node would consistently work after being added to the flow.

### Please tell us about your environment:
- [ ] Node-RED version: v0.18.7
- [ ] node.js version: v9.11.1
- [ ] Browser: Google Chrome Version 66.0.3359.181 
- [ ] npm version: 6.1.0
- [ ] Platform/OS: MacBook Pro: MacOS High Sierra Version 10.13.4 (17E202)

### Additional comments
I've developed 2 other custom nodes for Node-Red, using the same process & format, and I didn't have of the above-mentioned problems with those nodes. However, now I am seeing this issue with _several_ new nodes I'm making.  I even see this happen after completely re-writing the nodes.  Whatever is happening, it is consistent across 3 separate new projects - but non-existent on the older ones.  Hard to tell if this is a Node-Red issue or an issue with my code and/or development environment.  The thing that gives me the most pause is the inconsistent behavior without code alteration (otherwise I would 100% label this as a problem with my code).

**Comment by dceejay** [comment]
Does it happen if you use the recommended node 8.x LTS version ?
Can we see the code from one of the nodes that does this ?

**Comment by knolleary** [comment]
@computerFriend thanks for that.

The next time this happens, please run the following commands in the browser console, capture the output and send it to me via email or attach as a file here. 

    JSON.stringify(RED.nodes.registry.getModuleList())

    JSON.stringify(RED.nodes.registry.getNodeSetForType(NAME_OF_NODE_TYPE_YOU'RE_ADDING))

When you install/uninstall the node, are you doing so via the Palette Manager? Or on the command-line with a restart of node-red afterwards?

**Comment by computerFriend** [comment]
Hi Nick, Thanks for the feedback.
I'm installing using the command line (npm install/uninstall) & then manually removing any files/code that npm doesnt remove. Would it be better for me to use the Palette Manager?  (If so, how would I use the Palette Manager for local testing of a node in development)

@dceejay The code for the nodes I'm having problems with is all on github:
https://github.com/computerFriend/node-red-contrib-c8y-measuremts
https://github.com/computerFriend/node-red-contrib-c8y-alarms

Let me know if you have any more questions. I am going to run the above commands in the console & send you the output.

Thanks for your continued support!

**Comment by knolleary** [comment]
Have you looked at using `npm link` whilst developing your nodes? It will setup symlinks for you so you don't need to npm install/uninstall each time you want to update the nodes in node-red... you just need to restart node-red to pick up your latest changes.

For this issue, still with the command-line. Because you are restarting node-red each time, its starting from a known, clean point. When you install via the palette manager, its a slightly different code path for adding the newly installed module into the existing config (and also the palette manager only works for nodes published and installed from the public npm repo... so not suitable for local development).

**Comment by computerFriend** [comment]
Yes I use `npm link` sometimes.  Starting using npm install/uninstall as part of my debugging process for this issue ("sanity checks"; part of the test flow where I completely remove the modules from my computer & then pull down from github again)

**Comment by zeushammer** [comment]
Is there an update this issue? We might be able to provide more data if needed.

**Comment by knolleary** [comment]
@zeushammer no update from us on it. We've never reproduced it and as you can see, there's no update from @computerFriend on if they managed to solve it.

If you can reproduce this specific issue, then sharing details would be helpful.

**Comment by computerFriend** [comment]
Tbh this was probably some dark JS magic (e.g., some sort of invisible caching)
it happened quite a bit (over a year ago) - but I haven't touched this work since, because I quit that job & went back to video encoding where I belong lol

enjoyed working with Node-red though!

**Comment by zaphod-42** [comment]
I know this is closed, but in case anyone else stumbles here I had the same problem, and it turned out it was using a config option named "type" that broke everything, when looking at the event being emitted when the node was added, the "type" property was set to the default for my config option, rather than the type of node. I notice in the second repo linked above the same thing is happening. I changed my config option to "sensorType" and the issue was solved.
